### PR TITLE
CLI & Argparse

### LINKED ISSUE #14: App.py - CLI
Write the necessary code to allow the program to be run from the command line, from `app.py`.

- [x] [Argparse](https://docs.python.org/3/library/argparse.html#the-add-argument-method)
- [x] Make program work with passed arguments 

Notebooks:  
- https://github.com/datsoftlyngby/dat4sem2020spring-python/blob/master/notebooks/02-1b%20argparse.ipynb
- https://github.com/datsoftlyngby/dat4sem2020spring-python/blob/master/notebooks/02-1a%20CLI%20Programs.ipynb

### PR DESCRIPTION
**Pull request by Runi-VN**
- [x] Argparse
  - [x] Custom Argparse spacing. (Should probably just have used a library..)
  - [x] Custom types
- [x] App.py
- [x] Manual testing
  - [x] All arguments work from CLI
  - [x] All of program works from CLI
- [x] Cleanup

## Changes
- `get_tweet_data` has been changed to `handle_tweet_data` -> https://github.com/Hold-Krykke/PythonExam/pull/20/commits/7be50c6e5506887f3327e253b429e3cb43fabcb1
Best I could come up with, I swear. 

## Errors
This error occurs in some cases when plotting. I guess some of the data can't be transformed to certain plot types?
```
UserWarning: 'set_params()' not defined for locator of type <class 'pandas.plotting._matplotlib.converter.PandasAutoDateLocator'>
  ax.locator_params(integer=True)
```
## Current output
See comments

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,26 +1,116 @@
...
+def _restricted_float(val: float):
+    try:
+        val = float(val)
+    except ValueError:
+        raise argparse.ArgumentTypeError(
+            f"{val} not a floating-point literal")
+
+    if 0.0 < x > 1.0:
+        raise argparse.ArgumentTypeError(f"{val} not in range [0.0, 1.0]")
+    return val
```
Comment (MalteMagnussen) [inline app.py]: What happens here?

Comment (Runi-VN) [reply app.py]: So, on an argument you can pass a `type` such as `float`, which we need here. But it doesn't provide any checking besides if it can convert the value to float.

We also need to check if the value is within a range of floats ([0.0-1.0]) so it is easier to move the type check into its own function - and it is also the recommended way to handle types in an argparser.

On top of that we get custom error messages. Argparse' error messages are generally good enough, though.

I will be implementing custom types like this for most arguments.

In short:  
Try to convert argument to float, raise exception if not possible.  

if possible, check if within allowed range, raise exception if not.  

if both possible, return val as float

File: app.py [hunk 2]
```diff
@@ -1,26 +1,197 @@
...
+
+
+def _default_date():
+    """
+    Generates today and five days from now for use with default values of the date-argument
+    """
+    today = datetime.date.today()
+    five_days_from_now = today + datetime.timedelta(days=5)
+    # create readable format, as should be input
+    return ' '.join([today.strftime('%Y-%m-%d'), five_days_from_now.strftime('%Y-%m-%d')]) + ' (next 5 days)'
```
Comment (Runi-VN) [inline app.py]: This `(next 5 days)` should be changed, as it's most likely going to interfere with date-parsing at a later point in the program.

**Comment by Runi-VN** [comment]
## Current output by `app.py -h`
```bash
$ python app.py -h
usage: TweetScraper9000

        A program that scrapes Twitter for hashtags and performs a sentiment analysis on the results.
        Presents results in a chosen chart format.


positional arguments:
  hashtags              The hashtags to scrape.
                        EXAMPLE: 'trump biden'
                        -REQUIRED-


optional arguments:
  -h, --help            show this help message and exit

  -p PLOT_TYPE, --plot PLOT_TYPE
                        Plot chart type, choose one. VALUES=[bar, line, pie]
                         (default: pie)

  -l, --local           Pass to attempt scraping from local files.
                         (default: False)

  -s, --save            Pass to save plots locally
                        (if omitted will show plots instead)
                         (default: False)

  -c TWEET_COUNT, --count TWEET_COUNT
                        The amount of tweets to search for.
                         (default: 300)

  -d DATE DATE, --date DATE DATE
                        The date range (yyyy-mm-dd) to search for.
                        EXAMPLE: '2020-05-01 2020-05-05'.
                         (default: [datetime.date(2020, 5, 18), datetime.date(2020, 5, 23)])

  -se REMOVE_SENTIMENT, --sentiment REMOVE_SENTIMENT
                        Ignore specific sentiment.
                        VALUES=[Positive, Negative, Uncertain]
                         (default: None)

  -sh SEARCH_HASHTAGS [SEARCH_HASHTAGS ...]
                        Filter result data by specific hashtags.
                        EXAMPLE: '#Trump #Biden'
                         (default: None)

  -sm SEARCH_MENTIONS [SEARCH_MENTIONS ...]
                        Filter result data by specific mentions.
                        EXAMPLE: '@JoeBiden @folketinget'
                         (default: None)

  -su SEARCH_URLS [SEARCH_URLS ...]
                        Filter result data by specific URLs.
                        EXAMPLE: 'https://pic.twitter.com/'
                         (default: None)

  -cl ADVANCED: certainty low
                        The lower float value for determining if a sentiment is deemed uncertain.
                        EXAMPLE: 0.15.
                        VALUES: [0.00-1.00] (default: 0.25)

  -ch ADVANCED: certainty high
                        The higher float value for determining if a sentiment is deemed uncertain.
                        EXAMPLE: 0.65.
                        VALUES: [0.00-1.00] (default: 0.75)


Source: https://github.com/Hold-Krykke/PythonExam
Created by: Camilla, Malte, Asger, Rúni
```

File: app.py [hunk 3]
```diff
@@ -1,26 +1,403 @@
...
+    Generates today and five days from now for use with default values of the date-argument
+    """
+    today = datetime.now().date()
+    five_days_from_now = today + timedelta(days=5)
+    # create readable format, as should be input
+    # return [today.strftime('%Y-%m-%d'), five_days_from_now.strftime('%Y-%m-%d')]
+    return [today, five_days_from_now]
+
+
+def _filter_search_values(key: str, values: list, collection: list):
```
Comment (Runi-VN) [inline app.py:167]: Originally wanted to use get_by_key_value but python Filter does not allow handling many values:  
https://stackoverflow.com/a/4778869

File: app.py [hunk 4]
```diff
@@ -1,26 +1,403 @@
...
+    This item would be returned as it matches 2 of 3 values.
+
+    >>> list (original collection)
+    """
+    print(key, values, collection[:3])
+    return_data = []
+    for item in collection:
+        if any(val in values for val in item[key]):
+            return_data.append(item)
+    return return_data
```
Comment (Runi-VN) [inline app.py]: Originally wanted to use get_by_key_value but python Filter does not allow handling many values:  
https://stackoverflow.com/a/4778869

**Review by Castau — APPROVED** [review]

File: app.py [hunk 5]
```diff
@@ -1,26 +1,402 @@
...
+
+    def _split_lines(self, text, width):
+        # https://stackoverflow.com/a/29498894
+        # modify super to add newline between arguments
+        lines = super()._split_lines(text, width)
+        if text.endswith('\n'):  # custom newline
+            lines += ['']
+        else:
+            lines += ['']  # also add newline between different arguments
+        return lines
```
Comment (MalteMagnussen) [inline app.py:153]: What does this mean? 
What is a "formatter" in this context?

File: app.py [hunk 6]
```diff
@@ -1,26 +1,402 @@
...
+    values hold ['#trump', '#uselection'].
+
+    This item would be returned as it matches 2 of 3 values.
+
+    >>> list (original collection)
+    """
+    return_data = []
+    for item in collection:
+        if any(val in values for val in item[key]):
+            return_data.append(item)
```
Comment (MalteMagnussen) [inline app.py:185]: Nice! You figured it out :D

File: app.py [hunk 7]
```diff
@@ -1,26 +1,402 @@
...
-print('Generating presentations...')
-result = *analyzed_tweets  # do stuff
+def _restricted_float(val: float):
+    """
+    Only allow float values within our range [0.00-1.00]
+    """
+    try:
+        val = float(val)
+    except ValueError:
+        raise argparse.ArgumentTypeError(f"{val} not a floating-point literal")
```
Comment (MrHardcode) [inline app.py:24]: I still don't understand this. Is it needed when deciding the high and low % of sentiment?

File: app.py [hunk 8]
```diff
@@ -1,26 +1,402 @@
...
+    if mentions:
+        filtered_data = _filter_search_values(
+            'mentions', mentions, filtered_data)
+        print(f'Done filtering on mentions: {mentions}')
+    if urls:
+        filtered_data = _filter_search_values(
+            'tweet_urls', urls, filtered_data)
+        print(f'Done filtering on urls: {urls}')
+
+    return filtered_data
```
Comment (MalteMagnussen) [inline app.py:212]: So this is the "Mother" method for the helper methods?

File: app.py [hunk 9]
```diff
@@ -1,26 +1,402 @@
...
+    """
+    # prepare model for analysis
     train_model_if_necessary()
+    # Verify data
+    start_date, end_date = dates
+    if (start_date > end_date):
+        raise ValueError(f'Start date {start_date} may not be later than end date {end_date}')
+    # negate bool as to get meaning worthy of get_tweets
+    # (app asks, "do you want local search?" get_tweets asks, "do you want online search?")
+    fresh_search = not fresh_search
```
Comment (MalteMagnussen) [inline app.py:240]: Whats this?

File: app.py [hunk 10]
```diff
@@ -1,26 +1,402 @@
...
+        filtered_data = remove_sentiment(filtered_data, remove_sentiment)
+        print("Done removing sentiment...")
+
+    # Getting plot data from the get_sentiment function
+    plot_data = get_sentiment(filtered_data)
+    print("Done getting sentiment df for plotting...")
+
+    # Warn user that plt.show() is blocking
+    if not save_plot:
+        print('\tNOTICE\n\tShowing the plot will block the main thread.\n\tExit the plot display to continue program.')
```
Comment (MalteMagnussen) [inline app.py:271]: What is backslash t?

File: app.py [hunk 11]
```diff
@@ -1,26 +1,402 @@
...
+    print('--------------------------------------')
+    print('\tTOP 5 MENTIONS')
+    pprint(list(mention_stats.items())[:5], width=1)
+    print('--------------------------------------')
+    print('\t EXAMPLE OBJECT')
+    pprint(analyzed_tweets[0], width=1)
+    print('--------------------------------------')
+
+
+if __name__ == "__main__":
```
Comment (MalteMagnussen) [inline app.py:304]: So this is basically "if using CLI" ?

File: app.py (hunk 7 above)
Comment (Runi-VN) [reply app.py:24]: Yes, it is related to the `uncertainity_low` and `uncertainty_high` that can be passed to Sentiment Analysis.

Arguments can have types passed to them, specifically for this example (as seen in line 378, 385):

```py
parser.add_argument(
        '-cl',
        help="The lower float value for determining if a sentiment is deemed uncertain.\nEXAMPLE: 0.15.\nVALUES: [0.00-1.00]",
        type=_restricted_float,
        ...
```
What we could do here, is to pass `type=float` and argparse would try to parse the argument to a float. If it could not, it would throw an error.  
That is all good, and essentially what we are doing here - replicating the behavior of the argparse parser. The important part of the function is below the highlighted part, where we check if the float value is within the allowed range:

```py
if 0.0 < val > 1.0:
        raise argparse.ArgumentTypeError(f"{val} not in range [0.0, 1.0]")
    return val
```
Essentially we are taking the argparse parsers default method for parsing floats, and adding custom features in the form of:  
- Custom error messages if it couldn't parse
- Custom range-handling to only allow values that the program take [0.00-1.00]

File: app.py [hunk 12]
```diff
@@ -1,26 +1,402 @@
...
+    # turn Namespace object into usable dict
+    args_dict = vars(parser.parse_args())
+    print('--------------------------------------')
+    print('Received Values:')
+    pprint(args_dict, width=1)
+    print('--------------------------------------')
+
+    # extract items from dict
+    hashtags, plot_type, fresh_search, tweet_count, date, save_plot, remove_sentiment, search_hashtags, search_mentions, search_urls, certainty_low, certainty_high = itemgetter(
+        'hashtags', 'plot_type', 'fresh_search', 'tweet_count', 'date', 'save_plot', 'remove_sentiment', 'search_hashtags', 'search_mentions', 'search_urls', 'certainty_low', 'certainty_high')(args_dict)
```
Comment (MalteMagnussen) [inline app.py:399]: So CLI makes a dict, which you then destrucutre and feed to main method?

**Review by MalteMagnussen — APPROVED** [review]

File: app.py (hunk 6 above)
Comment (Runi-VN) [reply app.py:185]: Thanks! I really struggled with this, trying to convert it into a list comprehension. But it completely destroyed the functions readability, so I decided to write it out over a few lines instead. Probably for the best. 😄

File: app.py (hunk 5 above)
Comment (Runi-VN) [reply app.py:153]: The argparse constructor allows us to pass the following `formatter_class=argparse.HelpFormatter`.

The constructor only allows us to pass *one* formatter, despite several of the provided classes being helpful. This has been discussed and suggested to be changed several times, but argparse isn't really being updated. (probably why there are a lot of libraries that replaces/enhance argparse..)

As mentioned in this docstring, this simply allows us to use two formatters at the same time. It is very hackish.

On top of that, we modify the parser to split lines on every `\n` - something it doesn't do default (it is actually very hard to otherwise do a newline)

For descriptions of each formatter, see here:  
https://docs.python.org/3.8/library/argparse.html#formatter-class

**edit**:

It is being passed at line 308

File: app.py (hunk 10 above)
Comment (MrHardcode) [reply app.py:271]: \t == tab like \n == newline

File: app.py (hunk 8 above)
Comment (Runi-VN) [reply app.py:212]: Could say so, yeah. The flow goes:
Main  
prepare_data (which I consider the mother method, as it handles the different modules)  
_filter_data (which filters depending on arguments passed)

File: app.py (hunk 5 above)
Comment (MalteMagnussen) [reply app.py:153]: Alright. Thanks :) 
But what _is_ an argparse formatter? 
Is it "_just_" for formatting the help writeouts? 
image

File: app.py (hunk 9 above)
Comment (Runi-VN) [reply app.py:240]: Is it not clear from the documentation?

The program asks the user, *do you want to search locally or online?* (default False)

The webscraper asks the program, *do you want to search online or locally?* (default True)

We have to negate as they are opposite boolean values.

**This basically turns False to True or True to False**

Comment (MalteMagnussen) [reply app.py:240]: Yeah, I just meant "what does "not" do?" 
And its basically just a ! operator like we know from JavaScript. Thanks :)

File: app.py (hunk 10 above)
Comment (MalteMagnussen) [reply app.py:271]: Ahhhh. Nice. Thanks!

File: app.py (hunk 12 above)
Comment (Runi-VN) [reply app.py:399]: CLI actually makes a Namespace object which looks like this:  
`Namespace(bar='BAR', foo='1')`
It isn't always easy to grab these values, so the suggested way to grab them is as I have done here. A bit hard to read as we have so many arguments, but it works.

Comment (MalteMagnussen) [reply app.py:399]: https://docs.python.org/dev/library/argparse.html#argparse.Namespace

Perfect, thanks

File: app.py (hunk 11 above)
Comment (MrHardcode) [reply app.py:304]: Not necessarily just for CLI. If you e.g. start a Flask Server in a terminal this if-statement is also needed

Comment (MalteMagnussen) [reply app.py:304]: Ok, yeah, so its "If run from terminal" then? :)

Comment (Runi-VN) [reply app.py:304]: Here is a very good walkthrough of the definition and its other values:
https://stackoverflow.com/a/419185

Essentially, __name__ is "__main__" when the program is the main program. If it is called by something else, as a module would, it is typically not "__main__".

Maybe if you allowed Presentation.py to be run in two setups, either as a standalone, or as a module. When run as a module, you call each function and use it. But you can and should specify its execution order in that statement (*if __name__ == "__main__":*) for when it is called by itself, as the main program.
image

File: app.py (hunk 5 above)
Comment (Runi-VN) [reply app.py:153]: Yeah, you could say so. :)

The main issue is that one of the classes formats the *help* part for the program itself, while the others format the *help* part of the arguments.

We want both, so it looks readable.  
Compared to the output I wrote in the above comment, without formatters it would look *something like this*:

```
$ python app.py -h
usage: TweetScraper9000
A program that scrapes Twitter for hashtags and performs a sentiment analysis on the results. Presents results in a chosen chart format.

positional arguments:
  hashtags        The hashtags to scrape. EXAMPLE: 'trump biden' -REQUIRED-

optional arguments:
  -h, --help            show this help message and exit
  -p PLOT_TYPE, --plot PLOT_TYPE
                        Plot chart type, choose one. VALUES=[bar, line, pie] 
(default: pie)
  -l, --local           Pass to attempt scraping 
from local files. (default: False)
  -s, --save            Pass to save plots locally 
(if omitted will show plots instead) (default: False)
```
which might look ok-readable here, but definitely is not in a terminal!

**Review by MrHardcode — APPROVED** [review]
Okay very good

**Comment by Runi-VN** [comment]
> Okay very good

image
